### PR TITLE
[CI Environment] Diverse updates to deployments & removal

### PR DESCRIPTION
**Pull request by MrMCake**
# Description

- Added better error handling for deployments
- Reduced verbosity of test deployments (-> Resource Group creation)
- Updated removal order
- Updated parameter folder search with additional parameter (less 'magic strings')
- Updated naming in Test-ModuleLocally to latest schema
- Updated diverse ReadMes

| Pipeline |
| - |
| [AnalysisServices: Servers](https://github.com/Azure/ResourceModules/actions/workflows/ms.analysisservices.servers.yml) |

# Type of Change

Please delete options that are not relevant.

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Update to documentation

### CODE REVIEW

File: utilities/tools/Test-ModuleLocally.ps1 [hunk 1]
```diff
@@ -263,10 +263,10 @@ function Test-ModuleLocally {
...
                     $functionInput['retryLimit'] = 1 # Overwrite default of 3
                     # Loop through test parameter files
-                    foreach ($paramFilePath in $moduleParameterFiles) {
-                        Write-Verbose ('Deploy module [{0}] with parameter file [{1}]' -f $ModuleName, (Split-Path $paramFilePath -Leaf)) -Verbose
-                        if ($PSCmdlet.ShouldProcess(('Module [{0}] with parameter file [{1}]' -f $ModuleName, (Split-Path $paramFilePath -Leaf)), 'Deploy')) {
-                            New-TemplateDeployment @functionInput -ParameterFilePath $paramFilePath
+                    foreach ($moduleTestFile in $moduleTestFiles) {
+                        Write-Verbose ('Deploy module [{0}] with parameter file [{1}]' -f $ModuleName, (Split-Path $moduleTestFile -Leaf)) -Verbose
+                        if ($PSCmdlet.ShouldProcess(('Module [{0}] with parameter file [{1}]' -f $ModuleName, (Split-Path $moduleTestFile -Leaf)), 'Deploy')) {
+                            New-TemplateDeployment @functionInput -ParameterFilePath $moduleTestFile
```
Comment (eriqua) [inline utilities/tools/Test-ModuleLocally.ps1:269]: It looks like we renamed all parameter file occurrences in test occurrences. Should we also update the parameter name for this function (`ParameterFilePath`) ?

File: utilities/tools/Test-ModuleLocally.ps1 [hunk 2]
```diff
@@ -251,9 +251,9 @@ function Test-ModuleLocally {
                 # -----------------
                 if ($ValidationTest) {
                     # Loop through test parameter files
-                    foreach ($paramFilePath in $moduleParameterFiles) {
-                        Write-Verbose ('Validating module [{0}] with parameter file [{1}]' -f $ModuleName, (Split-Path $paramFilePath -Leaf)) -Verbose
-                        Test-TemplateDeployment @functionInput -ParameterFilePath $paramFilePath
+                    foreach ($moduleTestFile in $moduleTestFiles) {
+                        Write-Verbose ('Validating module [{0}] with parameter file [{1}]' -f $ModuleName, (Split-Path $moduleTestFile -Leaf)) -Verbose
```
Comment (eriqua) [inline utilities/tools/Test-ModuleLocally.ps1:255]:
```suggestion
                        Write-Verbose ('Validating module [{0}] with test [{1}]' -f $ModuleName, (Split-Path $moduleTestFile -Leaf)) -Verbose
```

File: utilities/tools/Test-ModuleLocally.ps1 [hunk 3]
```diff
@@ -263,10 +263,10 @@ function Test-ModuleLocally {
                 if ($DeploymentTest) {
                     $functionInput['retryLimit'] = 1 # Overwrite default of 3
                     # Loop through test parameter files
-                    foreach ($paramFilePath in $moduleParameterFiles) {
-                        Write-Verbose ('Deploy module [{0}] with parameter file [{1}]' -f $ModuleName, (Split-Path $paramFilePath -Leaf)) -Verbose
-                        if ($PSCmdlet.ShouldProcess(('Module [{0}] with parameter file [{1}]' -f $ModuleName, (Split-Path $paramFilePath -Leaf)), 'Deploy')) {
-                            New-TemplateDeployment @functionInput -ParameterFilePath $paramFilePath
+                    foreach ($moduleTestFile in $moduleTestFiles) {
+                        Write-Verbose ('Deploy module [{0}] with parameter file [{1}]' -f $ModuleName, (Split-Path $moduleTestFile -Leaf)) -Verbose
```
Comment (eriqua) [inline utilities/tools/Test-ModuleLocally.ps1:267]:
```suggestion
                        Write-Verbose ('Deploy module [{0}] with test [{1}]' -f $ModuleName, (Split-Path $moduleTestFile -Leaf)) -Verbose
```

File: utilities/tools/Test-ModuleLocally.ps1 [hunk 4]
```diff
@@ -263,10 +263,10 @@ function Test-ModuleLocally {
...
+                        if ($PSCmdlet.ShouldProcess(('Module [{0}] with parameter file [{1}]' -f $ModuleName, (Split-Path $moduleTestFile -Leaf)), 'Deploy')) {
```
Comment (eriqua) [inline utilities/tools/Test-ModuleLocally.ps1:268]:
```suggestion
                        if ($PSCmdlet.ShouldProcess(('Module [{0}] with test [{1}]' -f $ModuleName, (Split-Path $moduleTestFile -Leaf)), 'Deploy')) {
```

File: utilities/tools/Test-ModuleLocally.ps1 (hunk 1 above)
Comment (MrMCake) [reply utilities/tools/Test-ModuleLocally.ps1:269]: You mean re-name `-ParameterFilePath` to e.g. `-TestFilePath`? 

In this case I'd say now, as the function on it's own actually wants a ParameterFile (as per the New-Az*Deployment invocation). The reason we renamed it for our modules was, that there they are 'test files' with currently JSON-ARM (and soon Bicep).

On that note: That's the reason I did not rename anything about the dependency pipeline. It's still referencing parameter files - which they absolutly are in that case :)

File: utilities/pipelines/sharedScripts/Get-ModuleTestFileList.ps1 [hunk 5]
```diff
@@ -9,22 +9,33 @@ The relative path is returned instead of the full one to make paths easier to re
...
 .EXAMPLE
 Get-ModuleTestFileList -ModulePath 'C:\ResourceModules\arm\Microsoft.Compute\virtualMachines'
 
-Returns the relative file paths of all parameter files of the virtual machines module.
+Returns the relative file paths of all parameter files of the virtual machines module in folder '.test'.
+
+.EXAMPLE
+Get-ModuleTestFileList -ModulePath 'C:\ResourceModules\arm\Microsoft.Compute\virtualMachines' -SearchFolder 'parameters'
+
+Returns the relative file paths of all parameter files of the virtual machines module in folder 'parameters'.
```
Comment (eriqua) [inline utilities/pipelines/sharedScripts/Get-ModuleTestFileList.ps1:23]:
```suggestion
Returns the relative file paths of all test files of the virtual machines module in folder 'parameters'.
```

File: utilities/pipelines/sharedScripts/Get-ModuleTestFileList.ps1 [hunk 6]
```diff
@@ -9,22 +9,33 @@ The relative path is returned instead of the full one to make paths easier to re
...
 Mandatory. The module path to search in.
 
+.PARAMETER SearchFolder
+Optional. The folder to search for files in
+
 .EXAMPLE
 Get-ModuleTestFileList -ModulePath 'C:\ResourceModules\arm\Microsoft.Compute\virtualMachines'
 
-Returns the relative file paths of all parameter files of the virtual machines module.
+Returns the relative file paths of all parameter files of the virtual machines module in folder '.test'.
```
Comment (eriqua) [inline utilities/pipelines/sharedScripts/Get-ModuleTestFileList.ps1:18]:
```suggestion
Returns the relative file paths of all test files of the virtual machines module in the default test folder ('.test').
```

File: utilities/tools/Test-ModuleLocally.ps1 [hunk 7]
```diff
@@ -74,7 +74,7 @@ $TestModuleLocallyInput = @{
 }
 Test-ModuleLocally @TestModuleLocallyInput -Verbose
 
-Run all Pesters test for a given template and a Test-Az*Deployment using each parameter file in the module's parameter folder in combination with the template and the provided tokens
+Run all Pesters test for a given template and a Test-Az*Deployment using each test file in the module's .test folder in combination with the template and the provided tokens
```
Comment (eriqua) [inline utilities/tools/Test-ModuleLocally.ps1:77]:
```suggestion
Runs all Pester tests for a given template and a Test-Az*Deployment using each test file in the module's default test folder ('.test') in combination with the template and the provided tokens
```

**Review by eriqua — APPROVED** [review]
🆗

**Pull request merged**